source/extensions/priv/server/ntds_decrypt.c

```diff
@@ -0,0 +1,129 @@
+#include "precomp.h"
+
+void bytes_to_string(LPBYTE data, int length, LPSTR output){
+	for (int i = 0; i < length; i++){
+		sprintf(output + (i << 1), "%02X", ((LPBYTE)data)[i]);
+	}
+}
+
+BOOL decrypt_hash(encryptedHash *encryptedNTLM, decryptedPEK *pekDecrypted, char *hashString, DWORD rid){
+	BOOL cryptOK = FALSE;
+	BYTE encHashData[17] = { 0 };
+	BYTE decHash[17] = { 0 };
+
+	memcpy(&encHashData, &encryptedNTLM->encryptedHash, 16);
+	cryptOK = decrypt_rc4(pekDecrypted->pekKey, encryptedNTLM->keyMaterial, encHashData, 1, 16);
+	if (!cryptOK){
+		return FALSE;
+	}
+	cryptOK = decrypt_hash_from_rid(encHashData, &rid, decHash);
+	if (!cryptOK){
+		return FALSE;
+	}
+	bytes_to_string(decHash, 16, hashString);
+	return TRUE;
+}
+
+BOOL decrypt_hash_from_rid(LPBYTE encodedHash, LPDWORD rid, LPBYTE decodedHash){
+	typedef NTSTATUS(__stdcall *PSYS25)(IN LPCBYTE data, IN LPDWORD key, OUT LPBYTE output);
+	HMODULE hAdvapi = LoadLibrary("advapi32.dll");
+	if (hAdvapi == NULL){
+		return FALSE;
+	}
+	PSYS25 decryptFromRID = (PSYS25)GetProcAddress(hAdvapi, "SystemFunction025");
+	if (decryptFromRID(encodedHash, rid, decodedHash) != 0){
+		return FALSE;
+	}
+	return TRUE;
+}
+
+BOOL decrypt_hash_history(LPBYTE encHashHistory, size_t sizeHistory, decryptedPEK *pekDecrypted, DWORD rid, char *accountHistory, int *historyCount){
+	BOOL cryptOK = FALSE;
+	size_t sizeHistoryData = sizeHistory - 24;
+	int numHashes = (sizeHistoryData / 16);
+	memcpy(historyCount, &numHashes, sizeof(historyCount));
+	LPBYTE encHistoryData = (LPBYTE)malloc(sizeHistoryData);
+	LPBYTE decHistoryData = (LPBYTE)malloc((sizeHistoryData * 2));
+	memcpy(encHistoryData, encHashHistory + 24, sizeHistoryData);
+	cryptOK = decrypt_rc4(pekDecrypted->pekKey, encHashHistory + 8, encHistoryData, 1, sizeHistoryData);
+	if (!cryptOK){
+		return FALSE;

```

Comment:
Memory allocated for `encHistoryData` and `decHistoryData` is leaked if `decrypt_rc4` fails.